### ISSUE TITLE
Use True/False instead of 1/0 in OrderedDict popitem example

### ISSUE THREAD

**Issue by wanglc02** [python/cpython#107421]
# Documentation

https://docs.python.org/3.12/library/collections.html#ordereddict-examples-and-recipes

The documentation implements a `TimeBoundedLRU` and a `MultiHitLRUCache` with `OrderedDict`. Both implementations contain `self.cache.popitem(0)` where `self.cache` is an `OrderedDict` instance. However, method `popitem()` actually takes a boolean argument indicating whether to pop from the left end or the right end of the `OrderedDict` instance. Though `0` means `False`, using `0` will mislead readers into passing an *arbitrary* integer as an index to this method to pop the corresponding item from the OrderedDict. Changing `0` to `False` improves clarity, and avoids complaints from a type checker. `self.requests.popitem(0)` in the implementation of `MultiHitLRUCache` has the same issue.
In addition, `self.requests.pop(args, None)` in the implementation of `MultiHitLRUCache` should be changed to `self.requests.pop(args)` since the statements preceding it have already guaranteed that `self.requests` has key args. This change will reduce readers' work to think about when `self.requests` would not have key args.

### Linked PRs
* gh-107613

**Comment by shailshouryya** [comment]
Additional potential issue:

The `TimeBoundedLRU` class indicates it is a `"LRU Cache that invalidates and refreshes old entries."` but the implementation is
```
    def __call__(self, *args):
        if args in self.cache:
            self.cache.move_to_end(args)
            timestamp, result = self.cache[args]
            if time() - timestamp <= self.maxage:
                return result
```

Shouldn't this actually be
```
    def __call__(self, *args):
        if args in self.cache:
            self.cache.move_to_end(args) # moves the entry to the end, but this does not update the time to live
            timestamp, result = self.cache[args]
            if time() - timestamp <= self.maxage:
                self.cache[args] = (time(), result) # actually refresh the entry
                return result
```

in order to actually `[invalidate] and [refresh the] old entries`?

I can also update the docs for this issue once we figure this part out.

**Comment by wanglc02** [comment]
Hello, shailshouryya! Thank you for comment this, but I think the line you added `self.cache[args] = (time(), result) # actually refresh the entry` will change the original purpose of this LRU cache. This line refreshes the timestamp of the entry and updates its time to live. However, it will not refresh the content of the entry, `result`. Therefore, the cached content will never be updated by a new call to func, if users frequently access this content. If the cached content is a webpage that will be outdated after one day, and every minute a user accesses the webpage, then the cached context will continuously be reused, and will never be refreshed by a call to func every day.

In conclusion, I think the purpose of `TimeBoundedLRU` is to update the content of the entry regularly. How often users access the entry will not delay or force the update of the content.

**Comment by shailshouryya** [comment]
Got it, thanks for clarifying this! I'll update the docs to explicitly use a boolean value with `.popitem(last=False)` instead of the current, implicitly converted boolean value through a falsy check with `.popitem(0)`.

<hr>

For anyone wondering why my suggestion of adding
```
                self.cache[args] = (time(), result) # actually refresh the entry
```

will change the functionality, run the following example
```
# current implementation
from collections import OrderedDict
from time import time, sleep

class TimeBoundedLRU:
    "LRU Cache that invalidates and refreshes old entries."
    def __init__(self, func, maxsize=128, maxage=30):
        self.cache = OrderedDict()      # { args : (timestamp, result)}
        self.func = func
        self.maxsize = maxsize
        self.maxage = maxage
    def __call__(self, *args):
        if args in self.cache:
            self.cache.move_to_end(args)
            timestamp, result = self.cache[args]
            if time() - timestamp <= self.maxage:
                return result
        result = self.func(*args)
        self.cache[args] = time(), result
        if len(self.cache) > self.maxsize:
            self.cache.popitem(0)
        return result


def current_thing(arg):
    return time()


cache = TimeBoundedLRU(current_thing, maxsize=128, maxage=10)
count = 0
while count < 30:
    # there should only be 3 distinct values, and the values update every 10 seconds
    print(f'At count {count:>2} (timestamp {time():<19}): {cache(1)}')
    sleep(1)
    count += 1


```

and then compare the results from above to the following example that includes my initially proposed change:
```
from collections import OrderedDict
from time import time, sleep

class TimeBoundedLRU:
    "LRU Cache that invalidates and refreshes old entries."
    def __init__(self, func, maxsize=128, maxage=30):
        self.cache = OrderedDict()      # { args : (timestamp, result)}
        self.func = func
        self.maxsize = maxsize
        self.maxage = maxage
    def __call__(self, *args):
        if args in self.cache:
            self.cache.move_to_end(args)
            timestamp, result = self.cache[args]
            if time() - timestamp <= self.maxage:
                self.cache[args] = (time(), result)           # actually refresh the entry
                return result
        result = self.func(*args)
        self.cache[args] = time(), result
        if len(self.cache) > self.maxsize:
            self.cache.popitem(0)
        return result


def current_thing(arg):
    return time()


cache = TimeBoundedLRU(current_thing, maxsize=128, maxage=10)
count = 0
while count < 30:
    # there is only 1 distinct value, and the value never updates
    # UNLESS the cache runs out of space and the entry is evicted
    # due to more entries being recently used
    # (which does not happen in this scenario since the cache has a
    # maxsize of 128 and only 1 entry is in the cache)
    print(f'At count {count:>2} (timestamp {time():<19}): {cache(1)}')
    sleep(1)
    count += 1


```

As we can see, my proposed change of adding
```
self.cache[args] = (time(), result) # actually refresh the entry
```
would block the cache from updating after `maxage` seconds unless the `TimeBoundedLRU` cache is called with more than `maxsize` different arguments within the `maxage` time period, since calling the `TimeBoundedLRU` instance with the same argument(s) again would simply refresh the expiry time - _**even**_ when the expiry time has already passed.

**Comment by rhettinger** [comment]
> Though 0 means False, using 0 will mislead readers into passing an arbitrary integer as an index to this method to pop the corresponding item from the OrderedDict. 

FWIW, the `0` was intentional.  This API was specifically designed to help users who already know the list API with `somelist.pop()` meaning "pop from the right" and `somelist.pop(0)` meaning "pop from the left".  However, people are now accustomed to thinking in terms of type signatures, so switching to `last=False` would be an improvement and would better match the main docs shown above the recipe.

> In addition, self.requests.pop(args, None) in the implementation of MultiHitLRUCache should be changed to self.requests.pop(args) since the statements preceding it have already guaranteed that self.requests has key args. 

This is a safety measure to reduce complications that arise with threading.  If the underlying data structure was `set`, we would choose `discard(k)` instead of `remove(k)`.  For a dict, the equivalent is `pop(k, None)`.
